[feature](cloud) pack small files to reduce remote storage request overhead

liaoxin01 opened the pull request:
Co-authored-by: Luwei <luwei@selectdb.com>

### What problem does this PR solve?

### Release note

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [x] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [x] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [x] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

liaoxin01 committed:
[cloud](merge-file) add update merge file info api (#5293)

luwei16 committed:
[feature](cloud) Implement mergeFileManager (#5422)

luwei16 committed:
[feature](cloud) Implement reference counting–based recycling (#5445)

liaoxin01 committed:
[feature](cloud) Implement merge file system (#5503)

luwei16 committed:
[feature](cloud) correct merge file recycling reference counts (#5479)

liaoxin01 committed:
[feature](cloud) Implement merge file system for inverted index

liaoxin01 committed:
fix format

liaoxin01 committed:
fix local storage

liaoxin01 committed:
set enable_write_index_searcher_cache flalse

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage `56.27% (359/638)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/aa6b133a6710301fb744d6e5a40474daac1113a9_aa6b133a6710301fb744d6e5a40474daac1113a9_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/aa6b133a6710301fb744d6e5a40474daac1113a9_aa6b133a6710301fb744d6e5a40474daac1113a9_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.90% (1690/2089) |
| Line Coverage     | 66.79% (29884/44745) |
| Region Coverage   | 67.54% (14933/22111) |
| Branch Coverage   | 57.72% (7937/13750) |

liaoxin01 committed:
fix add_into_searcher_cache

liaoxin01 committed:
don't merge file for schema change

liaoxin01 commented:
run cloud_p0

luwei16 committed:
[feature](cloud) Add multi-vault support for merge file (#5589)

Issue Number: close #xxx

<!--Describe your changes.-->

gavinchou commented on be/src/runtime/exec_env_init.cpp:
```diff
@@ -397,6 +398,13 @@ Status ExecEnv::_init(const std::vector<StorePath>& store_paths,
     _workload_sched_mgr->start(this);
 
+    // Initialize merge file manager
+    _merge_file_manager = io::MergeFileManager::instance();
```
do we even need this if not cloud mode

liaoxin01 committed:
add debug log for merge reader

liaoxin01 commented:
run compile

dataroaring commented on regression-test/pipeline/cloud_p0/conf/be_custom.conf:
```diff
@@ -39,7 +39,7 @@ pipeline_task_leakage_detect_period_sec=1
 enable_table_size_correctness_check=true
 enable_brpc_connection_check=true
-enable_write_index_searcher_cache=true
+enable_write_index_searcher_cache=false
```
Why did you change the config item?

dataroaring commented on gensrc/proto/olap_file.proto:
```diff
@@ -151,6 +157,7 @@ message RowsetMetaPB {
 
     optional int64 visible_ts_ms = 1010;
     optional string reference_instance_id = 1011;  // For cluster snapshot
+    map<string, MergeFileSegmentIndexPB> merge_file_segment_index = 1012;
```
Pls add comments for k,v of the map.

dataroaring commented on be/src/olap/rowset/segment_v2/index_file_writer.cpp:
```diff
@@ -128,6 +129,11 @@ Status IndexFileWriter::add_into_searcher_cache() {
         }
+        if (dynamic_cast<io::MergeFileWriter*>(_idx_v2_writer.get()) != nullptr) {
+            // MergeFileWriter: file may be merged, skip cache for now.
+            // The cache will be populated on first read.
```
Bad performance is introduced?

dataroaring commented on be/src/io/fs/merge_file_manager.h:
```diff
@@ -0,0 +1,155 @@
+    Status init();
+
+    // Write a small file to the current merge file
+    Status append(const std::string& path, const Slice& data);
```
append_small_file() is a better name.

dataroaring commented on gensrc/proto/cloud.proto:
```diff
@@ -2122,6 +2122,43 @@ message CloneInstanceResponse {
+}
+
+message MergedFileInfoPB {
+    enum MergeFileState {
```
State instead of MergeFileState.

dataroaring commented on gensrc/proto/cloud.proto:
```diff
@@ -2122,6 +2122,43 @@ message CloneInstanceResponse {
+    }
+    optional int64 ref_cnt = 1;
+    optional int64 total_file_num = 2;
+    optional int64 left_file_num = 3;
```
left -> remaining.

dataroaring commented on be/src/io/fs/merge_file_manager.h:
```diff
@@ -0,0 +1,155 @@
+
+namespace doris::io {
+
+struct MergeFileSegmentIndex {
```
MergedFileLocation?

dataroaring commented on be/src/io/fs/merge_file_writer.cpp:
```diff
@@ -0,0 +1,201 @@
+    } else {
+        // Buffer small file data
+        for (size_t i = 0; i < data_cnt; ++i) {
+            _buffer.append(data[i].data, data[i].size);
```
Add a metrics for memory capacity used by _buffer.

dataroaring commented on be/src/io/fs/merge_file_writer.cpp:
```diff
@@ -0,0 +1,201 @@
+
+Status MergeFileWriter::_close_async() {
+    // If merge file is disabled, just use direct write
+    if (!config::enable_merge_file) {
```
Here we should follow _is_direct_write.

dataroaring commented on be/src/io/fs/merge_file_writer.cpp:
```diff
@@ -0,0 +1,201 @@
+
+Status MergeFileWriter::_close_sync() {
+    // If merge file is disabled, just use direct write
+    if (!config::enable_merge_file) {
```
ditto.

dataroaring commented on be/src/io/fs/merge_file_writer.cpp:
```diff
@@ -0,0 +1,201 @@
+    // Check if we should switch to direct write mode
+    if (!_is_direct_write && _bytes_appended + total_size > config::small_file_threshold_bytes) {
+        RETURN_IF_ERROR(_switch_to_direct_write());
+        _is_direct_write = true;
```
What if there are already some snall files merged?

dataroaring commented on be/src/cloud/config.h:
```diff
@@ -171,6 +171,15 @@ DECLARE_mInt64(warm_up_rowset_sync_wait_max_timeout_ms);
 DECLARE_mBool(enable_warmup_immediately_on_new_rowset);
 
+// Merge file manager config
+DECLARE_mBool(enable_merge_file);
```
enable_pack_small_file?

dataroaring commented on be/src/io/fs/merge_file_manager.cpp:
```diff
@@ -0,0 +1,511 @@
+        while (it != _global_index_map.end()) {
+            const auto& index = it->second;
+            if (active_merge_files.find(index.merge_file_path) == active_merge_files.end()) {
+                it = _global_index_map.erase(it);
```
We'd better clear immediately for normal path.

liaoxin01 committed:
fix merge file size

liaoxin01 commented:
run buildall

luwei16 committed:
[feature](cloud) enforce txn metadata and recycler safety

luwei16 committed:
fix code style

luwei16 commented:
run buildall

liaoxin01 committed:
fix partitial update

luwei16 committed:
fix close check

luwei16 committed:
Revert "fix close check"

This reverts commit 2cfe0b57e5f04cce9dfc3972713fcf52547dca63.

luwei16 committed:
fix merge file writer close check

liaoxin01 committed:
merge filesystem support exist and file size

liaoxin01 committed:
revert regression conf

liaoxin01 commented:
run buildall

luwei16 committed:
add merge file metrics and guard V1 inverted index tablets

expose bvar counters/averages in MergeFileManager, log detailed uploads, and add coverage
let MergeFileSystem implement exists/file_size with explicit index-map init tracking
skip merge-file wrapping when a tablet uses V1 inverted indexes to avoid missing metadata

luwei16 commented:
run buildall

luwei16 committed:
fix compile error

luwei16 commented:
run p0

luwei16 commented:
run cloud_p0

luwei16 committed:
parallel upload

luwei16 commented:
run buildall

luwei16 committed:
fix parallel put

liaoxin01 committed:
fix empty inverted index

liaoxin01 committed:
disable merge for hdfs

liaoxin01 commented:
run buildall

luwei16 committed:
fix be ut and add concurrent ut

luwei16 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage `56.29% (483/858)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/cdf45fb757fa55da778a4134ffb973a578736a06_cdf45fb757fa55da778a4134ffb973a578736a06_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/cdf45fb757fa55da778a4134ffb973a578736a06_cdf45fb757fa55da778a4134ffb973a578736a06_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 81.02% (1725/2129) |
| Line Coverage     | 66.67% (30411/45616) |
| Region Coverage   | 67.36% (15221/22597) |
| Branch Coverage   | 57.48% (8086/14068) |

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_cdf45fb757fa55da778a4134ffb973a578736a06/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_cdf45fb757fa55da778a4134ffb973a578736a06/fe_report/index.html)

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.82% (808/964)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_cdf45fb757fa55da778a4134ffb973a578736a06_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_cdf45fb757fa55da778a4134ffb973a578736a06_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.53% (24432/34156) |
| Line Coverage     | 58.05% (254223/437907) |
| Region Coverage   | 53.39% (212548/398097) |
| Branch Coverage   | 54.73% (90614/165552) |

liaoxin01 committed:
add concurrent load case

liaoxin01 committed:
add group commit case

liaoxin01 committed:
add mix load case

liaoxin01 committed:
add file cache clean case

liaoxin01 committed:
add test_merge_file_query_with_table_drop

liaoxin01 committed:
add ut

luwei16 committed:
improve merge

liaoxin01 committed:
fix case

liaoxin01 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage `56.29% (483/858)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/a2e59bacc178449a73d035592755425d20762019_a2e59bacc178449a73d035592755425d20762019_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/a2e59bacc178449a73d035592755425d20762019_a2e59bacc178449a73d035592755425d20762019_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 81.02% (1725/2129) |
| Line Coverage     | 66.65% (30403/45616) |
| Region Coverage   | 67.34% (15216/22597) |
| Branch Coverage   | 57.46% (8083/14068) |

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `75.03% (745/993)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/a2e59bacc178449a73d035592755425d20762019_a2e59bacc178449a73d035592755425d20762019/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/a2e59bacc178449a73d035592755425d20762019_a2e59bacc178449a73d035592755425d20762019/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.76% (18370/34821) |
| Line Coverage     | 38.21% (167357/438045) |
| Region Coverage   | 33.16% (130252/392856) |
| Branch Coverage   | 33.92% (55829/164572) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `87.32% (840/962)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_a2e59bacc178449a73d035592755425d20762019_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_a2e59bacc178449a73d035592755425d20762019_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.60% (24466/34169) |
| Line Coverage     | 58.11% (254615/438154) |
| Region Coverage   | 53.49% (212974/398124) |
| Branch Coverage   | 54.78% (90694/165565) |

luwei16 committed:
improve bvar

luwei16 committed:
limit small files num

luwei16 commented:
run buildall

liaoxin01 committed:
add checker

luwei16 committed:
fix merge file url

luwei16 committed:
fix be ut

luwei16 committed:
fix code style

luwei16 committed:
fix recycle merge file when drop table

starocean999 committed:
[fix](nereids)StatementContext must be created before creating stream load plan

starocean999 committed:
fix case

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `87.42% (848/970)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_2d14d7ec4bff2b09f4ff08b9656d209ab6677f80_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_2d14d7ec4bff2b09f4ff08b9656d209ab6677f80_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.68% (24643/34378) |
| Line Coverage     | 58.20% (257130/441769) |
| Region Coverage   | 53.31% (214392/402139) |
| Branch Coverage   | 54.80% (91598/167149) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `87.42% (848/970)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_2d14d7ec4bff2b09f4ff08b9656d209ab6677f80_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_2d14d7ec4bff2b09f4ff08b9656d209ab6677f80_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.69% (24645/34378) |
| Line Coverage     | 58.21% (257173/441769) |
| Region Coverage   | 53.33% (214443/402139) |
| Branch Coverage   | 54.81% (91617/167149) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `87.42% (848/970)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_f0e5a0004229fa39eed731c71e9d6facae523087_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_f0e5a0004229fa39eed731c71e9d6facae523087_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.72% (24657/34378) |
| Line Coverage     | 58.30% (257555/441769) |
| Region Coverage   | 53.40% (214725/402139) |
| Branch Coverage   | 54.93% (91820/167149) |

liaoxin01 committed:
fix checker

luwei16 committed:
fix

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage `43.11% (494/1146)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/483080d29bbccbbc03bbc6aa5ca88fcc50eebe22_483080d29bbccbbc03bbc6aa5ca88fcc50eebe22_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/483080d29bbccbbc03bbc6aa5ca88fcc50eebe22_483080d29bbccbbc03bbc6aa5ca88fcc50eebe22_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.81% (1727/2164) |
| Line Coverage     | 65.94% (30537/46313) |
| Region Coverage   | 66.59% (15238/22883) |
| Branch Coverage   | 56.85% (8101/14250) |

liaoxin01 committed:
rowset writer see different state when enable_merge_file changed

liaoxin01 commented:
run buildall

luwei16 committed:
add comment

luwei16 committed:
fix comment

luwei16 commented:
run buildall

luwei16 committed:
fix compile

liaoxin01 committed:
add bvar and fix cases

liaoxin01 commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_64cdd90e1a3988b7fc7cb747e1b97f6963db918f/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_64cdd90e1a3988b7fc7cb747e1b97f6963db918f/fe_report/index.html)

luwei16 committed:
fix clout ut

luwei16 commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57770_ab695fc167026cc855f12129497813c8875ff4cc/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57770_ab695fc167026cc855f12129497813c8875ff4cc/fe_report/index.html)

liaoxin01 committed:
Revert "add bvar and fix cases"

This reverts commit 64cdd90e1a3988b7fc7cb747e1b97f6963db918f.
